v0/salesforce/transform.js

```diff
@@ -0,0 +1,191 @@
+    rawPayload[mappingJson[sourceKey]] = get(message, sourceKey);
+  });
+  // Remove keys with undefined values
+  const payload = removeUndefinedValues(rawPayload);
+
+  //Get custom params from destination config
+  let customParams = getParamsFromConfig(message, destination);
+  customParams = removeUndefinedValues(customParams);
+  //request configuration will be conditional
+  //POST for create, PATCH for update
+
+  const response = {
+    endpoint: targetEndpoint,
+    requestConfig: defaultPostRequestConfig,
+    header: {
+      "Content-Type" : "application/json",
+      "Authorization" : authorizationHeader
+    },
+    userId: message.anonymousId,
+    payload: { ...customParams, ...payload }
+  };
+  return response;
+};
+
+function getParamsFromConfig(message, destination) {
+  let params = {};
+
+  var obj = {};
+  // customMapping: [{from:<>, to: <>}] , structure of custom mapping
+  if (destination.Config.customMappings) {
+    destination.Config.customMappings.forEach(mapping => {
+      obj[mapping.from] = mapping.to;
+    });
+  }
+  let keys = Object.keys(obj);
+  keys.forEach(key => {
+    params[obj[key]] = get(message.properties, key);
+  });
+  return params;
+}
+
+
+
+
+//Function for handling identify events
+async function processIdentify(message,destination) {
+
+  //start with creation endpoint, update only if Lead does not exist
+  var targetEndpoint = "https://" 
+                      + destination.Config.instanceName 
+                      + ".salesforce.com"
+                      + "/services/data/v"
+                      + SF_API_VERSION
+                      + "/sobjects/Lead";
+
+  //Get the authorization header if not available
+  if (!authorizationHeader) {
+    authorizationHeader = await getSFDCHeader(destination);
+  }
+  
+  //check if the lead exists
+  //need to perform a parameterized search for this using email
+  var email = message.context.traits.email;
+  
+  var leadQueryUrl = "https://" 
+                      + destination.Config.instanceName 
+                      + ".salesforce.com"
+                      + "/services/data/v"
+                      + SF_API_VERSION
+                      + "/parameterizedSearch/?q="
+                      + email
+                      + "&sobject=Lead&Lead.fields=id"
+
+
+  var leadQueryResponse = await axios.get(leadQueryUrl, 
+                                        {headers: {
+                                          "Authorization" : authorizationHeader  
+                                          }
+                                        });
+
+  var retrievedLeadCount = leadQueryResponse.data.searchRecords.length;
+  //if count is greater than zero, it means that lead exists, then only update it
+  //else the original endpoint, which is the one for creation - can be used 
+  if(retrievedLeadCount > 0) {
+    targetEndpoint += "/"+leadQueryResponse.data.searchRecords[0].Id+"?_HttpMethod=PATCH";
+  }                                      
+
```

Comment:
If email is not found from the search, targetEndpoint won't have the lead Id. Should we return an error in this case?

Reply:
The only difference between endipoints for CREATE and UPDATE is that UPDATE has the Lead ID at the end. So what the code does is it sets the CREATE endpoint first. If Lead ID is found, then that gets appended (along with the PATCH directive since we don’t support PATCH yet)

Reply:
Got it. Cool